File: node/antchain_node/transaction.py
"""Account-model transactions: sender, recipient, amount, sender nonce
(replay protection), ECDSA signature. No fees, no UTXO -- kept minimal since
the point of this project is the consensus mechanism, not payment-layer
design."""
from __future__ import annotations

from dataclasses import dataclass

from .crypto import canonical_json, verify_signature, address_from_pubkey, pubkey_from_hex


@dataclass
class Transaction:
    sender: str
    sender_pubkey: str
    recipient: str
    amount: float
    nonce: int
    signature: str = ""

    def signing_payload(self) -> dict:
        return {
            "sender": self.sender,
            "recipient": self.recipient,
            "amount": self.amount,
            "nonce": self.nonce,
        }

    def signing_bytes(self) -> bytes:
        return canonical_json(self.signing_payload())

    @property
    def tx_hash(self) -> str:
        import hashlib
        payload = dict(self.signing_payload())
        payload["signature"] = self.signature
        return hashlib.sha256(canonical_json(payload)).hexdigest()

    def sign(self, wallet) -> None:
        self.signature = wallet.sign(self.signing_bytes())

    def verify(self) -> bool:
        if self.amount <= 0:
            return False
        try:
            if address_from_pubkey(pubkey_from_hex(self.sender_pubkey)) != self.sender:
                return False
        except ValueError:
            return False
        return verify_signature(self.sender_pubkey, self.signing_bytes(), self.signature)

    def to_dict(self) -> dict:
        d = self.signing_payload()
        d["sender_pubkey"] = self.sender_pubkey
        d["signature"] = self.signature
        return d

    @staticmethod
    def from_dict(d: dict) -> "Transaction":
        return Transaction(
            sender=d["sender"],
            sender_pubkey=d["sender_pubkey"],
            recipient=d["recipient"],
            amount=d["amount"],
            nonce=d["nonce"],
            signature=d.get("signature", ""),
        )

File: cpp_node/src/transaction.cpp
#include "transaction.hpp"

#include <stdexcept>

namespace antchain_node {

Json Transaction::signing_payload() const {
    Json j = Json::object();
    j.set("sender", Json::string(sender));
    j.set("recipient", Json::string(recipient));
    j.set("amount", Json::number(amount));
    j.set("nonce", Json::integer(static_cast<int64_t>(nonce)));
    return j;
}

std::string Transaction::signing_bytes() const {
    return signing_payload().dump_canonical();
}

std::string Transaction::tx_hash() const {
    Json payload = signing_payload();
    payload.set("signature", Json::string(signature));
    return blake2b256_hex(payload.dump_canonical());
}

void Transaction::sign(const Wallet& wallet) {
    signature = wallet.sign_hex(signing_bytes());
}

bool Transaction::verify() const {
    if (amount <= 0.0) return false;
    try {
        if (address_from_pubkey_hex(sender_pubkey) != sender) return false;
    } catch (const std::exception&) {
        return false;
    }
    return verify_signature(sender_pubkey, signing_bytes(), signature);
}

Json Transaction::to_json() const {
    Json j = signing_payload();
    j.set("sender_pubkey", Json::string(sender_pubkey));
    j.set("signature", Json::string(signature));
    return j;
}

Transaction Transaction::from_json(const Json& j) {
    Transaction tx;
    tx.sender = j.at("sender").as_string();
    tx.sender_pubkey = j.at("sender_pubkey").as_string();
    tx.recipient = j.at("recipient").as_string();
    tx.amount = j.at("amount").as_double();
    tx.nonce = static_cast<uint64_t>(j.at("nonce").as_int());
    static const Json kEmpty = Json::string("");
    tx.signature = j.get("signature", kEmpty).as_string();
    return tx;
}

} // namespace antchain_node
